core/src/main/java/com/linecorp/armeria/server/websocket/WebSocketService.java

```diff
@@ -0,0 +1,278 @@
+    // See https://datatracker.ietf.org/doc/html/rfc6455#section-11.3.3
+    private static String generateSecWebSocketAccept(String webSocketKey) {
+        final String acceptSeed = webSocketKey + WEBSOCKET_13_ACCEPT_GUID;
+        final byte[] sha1 = Hashing.sha1().hashBytes(acceptSeed.getBytes(StandardCharsets.US_ASCII)).asBytes();
+        return Base64.getEncoder().encodeToString(sha1);
+    }
+
+    private HttpResponse handleUpgradeRequest(ServiceRequestContext ctx, HttpRequest req,
+                                              ResponseHeaders responseHeaders) {
+        if (ctx.requestTimeoutMillis() < closeTimeoutMillis) {
+            ctx.setRequestTimeoutMillis(TimeoutMode.EXTEND, closeTimeoutMillis);
+        }
+
+        final HttpResponseWriter responseWriter = HttpResponse.streaming();
+        responseWriter.write(responseHeaders);
+        responseWriter.whenConsumed().thenRun(() -> {
+            final WebSocketCloseHandler webSocketCloseHandler =
+                    new WebSocketCloseHandler(ctx, responseWriter, closeTimeoutMillis);
+            final WebSocketFrameDecoder decoder = new WebSocketFrameDecoder(
+                    ctx, decoderConfig, responseWriter, encoder, webSocketCloseHandler, true);
+            final StreamMessage<WebSocketFrame> inboundFrames = req.decode(decoder, ctx.alloc());
+            try {
+                final WebSocket outboundFrames = handler.handle(ctx, new WebSocketWrapper(inboundFrames));
+                outboundFrames.subscribe(new WebSocketResponseSubscriber(ctx, responseWriter,
+                                                                         encoder, webSocketCloseHandler),
+                                         ctx.eventLoop(), SubscriptionOption.WITH_POOLED_OBJECTS);
+            } catch (Throwable cause) {
+                req.abort(cause);
+                responseWriter.close(cause);
+            }
+        });
+        return responseWriter;
+    }
+
+    /**
+     * Handles the HTTP/2 web socket handshake described in
+     * <a href="https://datatracker.ietf.org/doc/html/rfc8441">Bootstrapping WebSockets with HTTP/2</a>.
+     * These are examples of a request and the corresponding response:
+     *
+     * <p>Request:
+     * <pre>
+     * HEADERS + END_HEADERS
+     * :method = CONNECT
+     * :protocol = websocket
+     * :scheme = https
+     * :path = /chat
+     * :authority = server.example.com
+     * sec-websocket-protocol = chat, superchat
+     * sec-websocket-extensions = permessage-deflate
+     * sec-websocket-version = 13
+     * origin = http://www.example.com
+     * </pre>
+     *
+     * <p>Response:
+     * <pre>
+     * HEADERS + END_HEADERS
+     * :status = 200
+     * sec-websocket-protocol = chat
+     * </pre>
+     */
+    @Override
+    protected HttpResponse doConnect(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+        if (!ctx.sessionProtocol().isHttp2()) {
```

Comment:
Just in case we add HTTP/3 support in the future :wink: 
```suggestion
        if (ctx.sessionProtocol().isHttp1()) {
```